### PR TITLE
Generate ZIP file using GitHub Actions

### LINKED ISSUE #6623: Generate a zip link on each PR
With the effort that @nielslange did in this package https://github.com/nielslange/woo-test-environment/ it would be nice to have an action that builds the blocks from the current branch and post it as a comment. People can use that zip locally or some other automation can use to post directly to a public jurassic ninja.

### PR DESCRIPTION
**Pull request by nielslange**
Fixes #6623

#### Notes

This PR aims to generate a ZIP file of the current PR so that this (development) version can be installed using [the custom WP-CLI command `woo-test-environment`](https://github.com/nielslange/woo-test-environment).

#### User Facing Testing

1. Create a feature branch based on this PR.
2. Apply some changes to the codebase, e.g. adding a test file/folder to the root folder.
3. Create a PR for the new feature branch.
4. Spin up a test site using https://jurassic.ninja/.
5. Install the custom WP-CLI command woo-test-environment using the following command:
```sh
wp package install git@github.com:nielslange/woo-test-environment.git
```
6. Spin up a WooCommerce test environment with this PR using the following command:
```sh
wp woo-test-environment setup --blocks={ZIP-FILE-URL}
```
7. Look up the WooCommerce Blocks folder of the site.
8. Verify that the WooCommerce Blocks folder contains the file/folder as added in step 2.

* [x] Do not include in the Testing Notes

⚠️ Note: The URL for `{ZIP-FILE-URL}`, as mentioned in step 6. can be found in the comment of the PR as created in step 3.

### CODE REVIEW

**Comment by nielslange** [comment]
The release ZIP for this PR is accessible via https://nielslange.com/releases/woocommerce-gutenberg-products-block-6661.zip

**Comment by nielslange** [comment]
The release ZIP for this PR is accessible via:
```https://nielslange.com/releases/woocommerce-gutenberg-products-block-6661.zip```

**Comment by nielslange** [comment]
The release ZIP for this PR is accessible via:
```
https://nielslange.com/releases/woocommerce-gutenberg-products-block-6661.zip
```

**Comment by nielslange** [comment]
The release ZIP for this PR is accessible via:
```
https://wcblocks.wpcomstaging.com/wp-content/uploads/woocommerce-gutenberg-products-block-6661.zip
```

**Comment by nielslange** [comment]
The release ZIP for this PR is accessible via:
```
https://wcblocks.wpcomstaging.com/wp-content/uploads/woocommerce-gutenberg-products-block-6661.zip
```

**Comment by nielslange** [comment]
The release ZIP for this PR is accessible via:
```
https://wcblocks.wpcomstaging.com/wp-content/uploads/woocommerce-gutenberg-products-block-6661.zip
```

**Comment by nielslange** [comment]
The release ZIP for this PR is accessible via:
```
https://wcblocks.wpcomstaging.com/wp-content/uploads/woocommerce-gutenberg-products-block-6661.zip
```

File: .github/workflows/generate-zip.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+          port: 22
+          username: ${{ secrets.FTP_USER }}
+          password: ${{ secrets.FTP_PASS }}
+          sourceDir: ./temp/
+          targetDir: ${{ secrets.FTP_DIR }}
+
+      - name: Add release ZIP URL as comment to the PR
+        uses: mshick/add-pr-comment@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.API_TOKEN_GITHUB }}
```
Comment (tarunvijwani) [inline .github/workflows/generate-zip.yml]: At the moment, it adds comments from your account. I think that we should add the comment using the [Rubik User Bot](https://github.com/rubikuserbot) account. We can use the same GitHub token that is used here: https://github.com/woocommerce/woocommerce-blocks/blob/trunk/.github/workflows/automate-team-review-assignment-config.yml#L15

```suggestion
          GITHUB_TOKEN: ${{ secrets.PAT_FOR_ACTIONS }}
```

**Review by tarunvijwani — CHANGES_REQUESTED** [review]
Thank you for working on this @nielslange. It works like a charm.

File: .github/workflows/generate-zip.yml (hunk 1 above)
Comment (nielslange) [reply .github/workflows/generate-zip.yml]: Thanks for pointing this out, @tarunvijwani. I just committed your suggestion.

File: .github/workflows/generate-zip.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+      - name: Add release ZIP URL as comment to the PR
+        uses: mshick/add-pr-comment@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.PAT_FOR_ACTIONS }}
+        with:
+          message: |
+            The release ZIP for this PR is accessible via:
+            ```
+            https://wcblocks.wpcomstaging.com/wp-content/uploads/woocommerce-gutenberg-products-block-${{ github.event.pull_request.number }}.zip
```
Comment (sunyatasattva) [inline .github/workflows/generate-zip.yml:58]: Is there a specific reason why are we uploading the zip under the uploads for a WordPress installation? What about just creating an independent dir at the root level?

File: .github/workflows/generate-zip.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+      - name: Install Node dependencies
+        if: steps.cache-node-modules.outputs.cache-hit != 'true'
+        run: npm ci --no-optional
+
+      - name: Generate ZIP file
+        run: npm run package-plugin:deploy
+
+      - name: Create temp folder
+        run: mkdir temp
```
Comment (sunyatasattva) [inline .github/workflows/generate-zip.yml]: Just for the sake of looking long-term, what about naming this `temp` folder to something that's less likely to cause any conflicts in the future? Perhaps namespacing it like: `wc-blocks-pr-release__temp`.

**Review by sunyatasattva — APPROVED** [review]
Looks great, thank you @nielslange ! I've just got some tiny comments, but nothing blocking, so I'm pre-approving.

File: .github/workflows/generate-zip.yml (hunk 2 above)
Comment (nielslange) [reply .github/workflows/generate-zip.yml:58]: > Is there a specific reason why are we uploading the zip under the uploads for a WordPress installation? What about just creating an independent dir at the root level?

I've decided to upload the ZIP files to a WordPress.com site. The site is part of a multisite and AFAIK, it's not possible to access the root directly itself. WordPress only allows public access to the uploads folder. That's why I decided to transfer the ZIP files to this folder.

File: .github/workflows/generate-zip.yml (hunk 3 above)
Comment (nielslange) [reply .github/workflows/generate-zip.yml]: Great suggestion. I just changed it!

**Comment by nielslange** [comment]
Thanks for your reviews, @tarunvijwani and @sunyatasattva! 🙌

File: .github/workflows/generate-zip.yml (hunk 2 above)
Comment (sunyatasattva) [reply .github/workflows/generate-zip.yml:58]: I see, thanks for the answer. Is there a way to maybe then group those uploads under a subdir of `uploads` (like `uploads/wc-blocks-pr-releases` or something) or would that mess with WordPress to have an unknown subdir within `uploads`?

That's just out of curiosity.

**Comment by dinhtungdu** [comment]
@nielslange Do you think we should create only one comment for the release ZIP file link? As the comments are the same, I don't think we need to create more than one per PR. We're doing a similar thing to the compare action at #6736.

**Comment by nielslange** [comment]
> @nielslange Do you think we should create only one comment for the release ZIP file link? As the comments are the same, I don't think we need to create more than one per PR. We're doing a similar thing to the compare action at #6736.

Technically, we could create only one comment. But we would then have to search the corresponding comment. When adding a comment after every push, the link to the ZIP file can always be found as the last comment, which increases the usability. 

Feel free to create an issue and a PR, if you believe that we benefit from having only one comment. I, personally, don't see an improvement in having only one comment and being forced to search it.

**Comment by jeffpaul** [comment]
@nielslange @dinhtungdu not sure if it would be possible, but what about adding that link at the bottom of the PR description in a new section, like:
> The release ZIP for this PR is accessible via: https://wcblocks.wpcomstaging.com/wp-content/uploads/woocommerce-gutenberg-products-block-6661.zip 

Or perhaps using a custom text field in the GitHub project board and adding the link into that field as it would be always available in the top right sidebar on GitHub?

**Comment by dinhtungdu** [comment]
In terms of usability, there is an issue I have with the current behavior of this action and assets compare action: they create additional notifications in the inbox.



> Or perhaps using a custom text field in the GitHub project board and adding the link into that field as it would be always available in the top right sidebar on GitHub?

If we can do this, it's the best way IMO.

**Comment by nielslange** [comment]
> @nielslange @dinhtungdu not sure if it would be possible, but what about adding that link at the bottom of the PR description in a new section [...]
> 
> Or perhaps using a custom text field in the GitHub project board and adding the link into that field as it would be always available in the top right sidebar on GitHub?

@jeffpaul I like both ideas, but couldn't find any information on doing that. While it might not be the perfect solution, it's a working solution. Given the amount of other tasks on my plate, I would rather not invest too much time in this issue. But if you find any information on how to edit a PR description or how to add additional sidebar sections using GitHub Actions, I'm all ears.

> In terms of usability, there is an issue I have with the current behavior of this action and assets compare action: they create additional notifications in the inbox.

@dinhtungdu Correct me if I'm wrong, but aren't notifications a different topic? Every GitHub Action run triggers a notification. Even if we would add the ZIP link to the PR description or in a sidebar panel, you might still receive a notification that the GitHub Action ran. That said, if it really bothers you, feel free to create a PR that improves the current behaviour. I just don't see how having a different behaviour would affect the notifications.

**Comment by dinhtungdu** [comment]
@nielslange I should've explained my point more clearly.

To give you more context, I turned off the notification for Actions, I also turn off the email notification. The only source of notifications I'm using is the GitHub Notifications Inbox. When the action creates a comment on a PR, it also triggers a new notification in the inbox which is false-positive to me. 

FWIW, I strongly agree that the benefit of "the link to the ZIP file can always be found as the last comment" outperforms other cons. So I don't think my original [comment](https://github.com/woocommerce/woocommerce-blocks/pull/6661#issuecomment-1193708433) is worth implementing anymore.

**Comment by nielslange** [comment]
Thanks for explaining the case you're facing, @dinhtungdu. A quick fix, I can think of, would be to change `allow-repeats: true` to `allow-repeats: false` in

https://github.com/woocommerce/woocommerce-blocks/blob/995dc562f5b6d51c3de2d11872c0d5e6d6fdd111/.github/workflows/generate-zip.yml#L50-L60

This definition causes the comment to be added when creating the PR and on every push. Setting the definition to `allow-repeats: false` will only trigger the first comment. Do you want to give this change a try to see if that approach works better for you?

**Comment by senadir** [comment]
> Technically, we could create only one comment. But we would then have to search the corresponding comment. When adding a comment after every push, the link to the ZIP file can always be found as the last comment, which increases the usability.

I agree with @dinhtungdu, the benefit of having a single comment reused is much greater than the effort required to find it, given most PRs won't get past 5 comments at best. The downside of a new comment on each commit is just too much noise to the point that you won't checking notification anymore.

**Comment by nielslange** [comment]
@jeffpaul and @dinhtungdu To closing the loop, we created two changes:

- https://github.com/woocommerce/woocommerce-blocks/pull/6755 → This PR prevents triggering email notifications.
- https://github.com/woocommerce/woocommerce-blocks/pull/6769 → This PR ensures that the comment with the link will only be added ones.

**Pull request merged**